src/java/org/apache/cassandra/transport/QueueBackpressure.java

```diff
@@ -0,0 +1,148 @@
+ * <p>
+ * If we have not seen old requests in the head of the queue for 1 second, we close the incident.
+ * <p>
+ * If the queue remains saturated for a prolonged period (meaning {@link Dispatcher#hasQueueCapacity} returns false), the amount of delay
+ * will increase in proportion to the request rate as appliedTimes & severityLevel are incremented. If no new requests are considered
+ * overloaded in this way for a second, the incident will be reset and so the delay will drop back down to {@link Impl#minDelayNanos()}.
+ */
+public interface QueueBackpressure
+{
+    QueueBackpressure NO_OP = timeUnit -> 0;
+
+    QueueBackpressure DEFAULT = new QueueBackpressure()
+    {
+        private final AtomicReference<Impl> state = new AtomicReference<>(noBackpressure(() -> DatabaseDescriptor.getNativeTransportMinBackoffOnQueueOverload(TimeUnit.NANOSECONDS),
+                                                                                         () -> DatabaseDescriptor.getNativeTransportMaxBackoffOnQueueOverload(TimeUnit.NANOSECONDS)));
+
+        public long markAndGetDelay(TimeUnit timeUnit)
+        {
+            return state.updateAndGet(Impl::mark).delay(timeUnit);
+        }
+    };
+
+    long markAndGetDelay(TimeUnit timeUnit);
+
+
+    static Impl noBackpressure(LongSupplier minDelayNanos, LongSupplier maxDelayNanos)
+    {
+        return new Impl(minDelayNanos, maxDelayNanos,
+                        -1, 0, 0);
+    }
+
+    @VisibleForTesting
+    class Impl
+    {
+        private final long appliedAt;
+
+        private final int severityLevel;
+        private final int appliedTimes;
+
+        private final LongSupplier minDelayNanos;
+        private final LongSupplier maxDelayNanos;
+
+        @VisibleForTesting
+        public Impl(LongSupplier minDelayNanos, LongSupplier maxDelayNanos, long appliedAt, int severityLevel, int appliedTimes)
+        {
+            this.minDelayNanos = minDelayNanos;
+            this.maxDelayNanos = maxDelayNanos;
+            this.appliedAt = appliedAt;
+            this.severityLevel = severityLevel;
+            this.appliedTimes = appliedTimes;
+        }
+
+
+        public Impl mark()
+        {
+            return mark(preciseTime.now());
+        }
+
+        @VisibleForTesting
+        public Impl mark(long now)
+        {
+            // Last time we have applied backpressure was over a second ago, consider this a new incident
+            if (appliedAt > 0 && now - appliedAt >= TimeUnit.SECONDS.toNanos(1))
+            {
+                return new Impl(minDelayNanos, maxDelayNanos, now, 1, 1);
+            }
+            // Continuing incident: apply backpressure but do not bump severity level yet
+            else if (appliedTimes < 10)
+            {
+                return new Impl(minDelayNanos, maxDelayNanos, now, severityLevel == 0 ? 1 : severityLevel, appliedTimes + 1);
```

Comment:
nit: Can we just start the `severityLevel` at 1?

Reply:
nit: It seems like there's a lot of `Impl` creation going on here during a spike. Is there any way we could perhaps moderate that a tiny bit by perhaps making `appliedTimes` an `AtomicInteger`?

Reply:
Unfortunately because we have at least two variables that we need to update atomically, now and appliedTimes, we will either have to create some sort of object, or do some binary math (but then we lose precision). I am afraid I could not find a quick and easy way to make this more lightweight.

I would also like to highlight that as soon as we have applied timeout, we will have the client back-off for the given amount of time, so this might be less of a problem: we do this only if there is no capacity in the queue.

Also, incident does start at 1, could you elaborate?

Reply:
I was just wondering why we had to do `severityLevel == 0 ? 1 : severityLevel` rather than just starting incidents at 1, but that means you have to check `appliedAt` or something in `delay()` to make sure you get zero before an incident has actually started.